src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c

```diff
@@ -179,6 +191,169 @@ xocl_ctx_to_info(struct drm_xocl_ctx *args, struct kds_ctx_info *info)
 		info->flags = CU_CTX_SHARED;
 }
 
+static int
+xocl_cu_ctx_to_info(struct xocl_dev *xdev, 
+		struct drm_xocl_hw_ctx *args, struct kds_ctx_info *info)
+{
+	uint32_t slot_hndl = args->hw_context; // Get the slot handler
+	struct kds_sched *kds = &XDEV(xdev)->kds;
+	char *kname_p = args->cu_name;
+	struct xrt_cu *xcu = NULL;
+	char iname[64];
+	char kname[64];
+	int i = 0;
+
+	strcpy(kname, strsep(&kname_p, ":"));
+	strcpy(iname, strsep(&kname_p, ":"));
+
+	/* Retrive the CU index from the given slot */
+	for (i = 0; i < MAX_CUS; i++) {
+		xcu = kds->cu_mgmt.xcus[i];
+		if (!xcu)
+			continue; 
+
+		if ((xcu->info.slot_idx == slot_hndl) &&
+				(!strcmp(xcu->info.kname, kname)) &&
+				(!strcmp(xcu->info.iname, iname))) {
+			info->cu_domain = DOMAIN_PL;
+			info->cu_idx = i;
+			goto done;
+		}
+	}
+
+	/* Retrive the CU index from the given slot */
+	for (i = 0; i < MAX_CUS; i++) {
+		xcu = kds->scu_mgmt.xcus[i];
+		if (!xcu)
+			continue; 
+
+		if ((xcu->info.slot_idx == slot_hndl) &&
+				(!strcmp(xcu->info.kname, kname)) &&
+				(!strcmp(xcu->info.iname, iname))) {
+			info->cu_domain = DOMAIN_PS;
+			info->cu_idx = i;
+			goto done;
+		}
+	}
+
+	return -EINVAL;
+
+done:
+	if (args->flags == XOCL_CTX_EXCLUSIVE)
+		info->flags = CU_CTX_EXCLUSIVE;
+	else
+		info->flags = CU_CTX_SHARED;
+
+	return 0;
+}
+
+static int xocl_open_cu_context(struct xocl_dev *xdev, struct kds_client *client,
+			    struct drm_xocl_hw_ctx *args)
+{
+	struct kds_ctx_info	 info;
+	xuid_t *uuid;
+	int ret;
+
+	mutex_lock(&client->lock);
+	/* If this client has no opened context, lock bitstream */
+	if (!client->ctx) {
+		/* Allocate the new client context and store the xclbin */
+		client->ctx = vzalloc(sizeof(struct kds_client_ctx));
+		if (!client->ctx) {
+			ret = -ENOMEM;
+			goto out;
+		}
+		
+		/* TODO : We need to use args->hw_context here to get uuid from a slot
+		 * for multi slot case.
+		 **/
+		ret = XOCL_GET_XCLBIN_ID(xdev, uuid);
+		if (ret) {
+			vfree(client->ctx);
+			client->ctx = NULL;
+			goto out;
+		}
+
+		client->ctx->xclbin_id = uuid;
+		XOCL_PUT_XCLBIN_ID(xdev);
+
+		client->ctx->slot_idx = args->hw_context;
+		list_add_tail(&client->ctx->link, &client->ctx_list);
+	}
```

Comment:
This is related to what I commented in kds_client.h. It looks like the client->ctx is per xclbin not per CU.

By the way, there is no lock bitstream step here. The comment before the if statement is incorrect.